### PR TITLE
(PE-33177) Update ClientOptions default SSL protocols

### PR DESCRIPTION
**Pull request by nmburgan**
### (PE-33177) Update ClientOptions default SSL protocols 
This updates the DEFAULT_SSL_PROTOCOLS variable to TLSv1.2 and TLSv1.3, to be in line with the desired defaults used in Puppet and PE.

### (PE-33177) Update exception handling in sync_ssl_test for clojure sync client
The clojure sync client now throws a ConnectionClosedException, rather than an SSLException, when a connection can't be established due to no valid protocols. This happens in the latest Java 11 versions, but older Java 11 versions (e.g. 11.0.2 used by Travis) still throws an SSLHandshakeException.  This commit changes java-unsupported-protocol-exception? to be able to handle both of these cases.

Note that there still seems to be cases where ConnectionClosedException is picked up by the HttpClientException block (using the Java sync client) and some cases where it isn't (Clojure sync client).

### (PE-33177) Update CHANGELOG and bump project version 
Creating a new major since this removes TLSv1 and TLSv1.1 from the defaults.

### (PE-33177) Update Travis to use more recent builds for java 8 and 11 
The versions of Java Travis has pre-installed are pretty old. The sync_ssl_test is looking for particular exceptions that have changed since these old versions. This changes the instructions for Travis to install OpenJDK from the Ubuntu repos, which are more up to date (though still not latest).

This also modifies the jobs to run in parallel, rather than sequentially, since they don't depend on each other. This wil save some time.

### CODE REVIEW

**Comment by nmburgan** [comment]
This is failing because Travis is using an OpenJDK8 version that is too old.  We need at least version 262, I think.

**Review by mwaggett — APPROVED** [review]
Do we have a plan for fixing Travis? A brief google search suggests it won't be the prettiest, but.. ¯\_(ツ)_/¯

**Comment by nmburgan** [comment]
I couldn't find how to update Travis' openjdk8 build.  What did you find?

**Comment by mwaggett** [comment]
something like this: https://stackoverflow.com/questions/29636754/can-you-specify-minor-jdk-version-for-travis-ci/30143945

I tried updating the `dist` in the Travis config (see https://github.com/puppetlabs/clj-http-client/pull/88), but bionic apparently uses an even older Java 8 build and focal apparently refuses to install Java 8 in the first place. 🙃

**Comment by justinstoller** [comment]
I think that's failing because something small changed in a java update and now this macro isn't quite right: https://github.com/puppetlabs/clj-http-client/blob/main/test/puppetlabs/http/client/sync_ssl_test.clj#L128-L143

File: CHANGELOG.md
Comment (mwaggett) [inline CHANGELOG.md:2]: Might want to also note the removal the old TLS versions.

**Review by mwaggett — APPROVED** [review]

File: test/puppetlabs/http/client/sync_ssl_test.clj
Comment (mwaggett) [inline test/puppetlabs/http/client/sync_ssl_test.clj:186]: oh fun, looks like Java 11 throws `SSLHandshakeException`

Comment (nmburgan) [reply test/puppetlabs/http/client/sync_ssl_test.clj:186]: Huh, it didn't locally for me.  Yay for Java flavor nuances :(

File: test/puppetlabs/http/client/sync_ssl_test.clj
Comment (nmburgan) [inline test/puppetlabs/http/client/sync_ssl_test.clj:144]: Not a huge fan of this, but the exception really doesn't contain any useful information to differentiate it from other reasons the connection might have been closed.

File: test/puppetlabs/http/client/sync_ssl_test.clj
Comment (nmburgan) [reply test/puppetlabs/http/client/sync_ssl_test.clj]: Appears to have changed between 11.0.2 and 11.0.13.

File: test/puppetlabs/http/client/sync_ssl_test.clj
Comment (mwaggett) [reply test/puppetlabs/http/client/sync_ssl_test.clj:144]: huh. so it doesn't work to just use `java-unsupported-protocol-exception?` in the test, without your addition?

Comment (nmburgan) [reply test/puppetlabs/http/client/sync_ssl_test.clj:144]: No.  Apparently it isn't an `HttpClientException` when using the Clojure sync client, for some reason.  But maybe sometimes it is with older Java versions?  I'm kind of confused about that, and I couldn't suss out exactly where that split happens.

File: .travis.yml
Comment (nmburgan) [reply .travis.yml]: Sorry, this was an old experiment.  Opted go with installing from the Ubuntu repo instead.  But if that becomes problematic, we could go back to using this script with a particular URL.

File: .travis.yml
Comment (nmburgan) [reply .travis.yml]: Old experiment, I removed this

File: .travis.yml
Comment (nmburgan) [reply .travis.yml]: Old experiment, I removed this

**Review by mwaggett — APPROVED** [review]
seems fine I guess, but will defer to @justinstoller

File: test/puppetlabs/http/client/sync_ssl_test.clj
Comment (justinstoller) [reply test/puppetlabs/http/client/sync_ssl_test.clj:144]: So, I looked into this a bit and raising a HttpClientException is part of our API which Puppet Server depends on (to catch and re-raise the exception that the Puppet client is expecting. However, it only looks like this test fails like this when using the clojure client, not the Java client that we use in Puppet Server. It also doesn't seem like it's too big of deal if on very rare occasions we break that api contract because the outer layer of the Puppet http client blanket catches exceptions and wraps them in the correct exception.

tl;dr I don't think it would break the API in Puppet Server and if it did for some reason it probably would have a negligible impact.

Having said that, I'm going to look at this for the last half hour of my day and see if I can see why it only fails in the clojure wrapper of the client.

Comment (justinstoller) [reply test/puppetlabs/http/client/sync_ssl_test.clj:144]: So, I did discover that this is behavior is how the Java async client works, while the Java sync client catches any errors and wraps them in the HttpClientException and rethrows. The clojure sync client is a thin wrapper around the clojure async client which is then a thin wrapper around the java async client. Leaving only the java sync client to implement the desired behavior. Since that's what Puppet Server uses I think we can be confident that this isn't masking some issue in Puppet.

I think _at some point_ it'd be nice to make the other APIs wrap exceptions in HttpClientExceptions, but I don't think we need to block this on it.

**Pull request merged**
